storage/pinlogs_blockwise.h

```diff
@@ -0,0 +1,91 @@
+
+
+#if FLASH_BLOCK_WORDS <= 1
+#error "FLASH_BLOCK_WORDS must be at least 2 to fit the counter and header"
+#endif
+
+#if PIN_MAX_TRIES > 16
+#error "PIN_MAX_TRIES must be at most 16"
+#endif
+
+#define PIN_LOG_BYTES ((FLASH_BLOCK_WORDS - 1) * sizeof(uint32_t))
+
+static uint8_t expand_counter(uint8_t c) {
+  c = ((c << 2) | c) & 0x33;
+  c = ((c << 1) | c) & 0x55;
+  c = ((c << 1) | c) ^ 0xaa;
+  return c;
+}
+
+static uint8_t compress_counter(uint8_t c) {
+  if (((c ^ (c << 1)) & 0xAA) != 0xAA) {
+    handle_fault("ill-formed counter");
+  }
+  c = c & 0x55;
+  c = ((c >> 1) | c) & 0x33;
+  c = ((c >> 2) | c) & 0x0f;
+  return c;
+}
+
+static secbool pin_get_fails(uint32_t *ctr) {
+  const void *logs = NULL;
+  uint16_t len = 0;
+
+  wait_random();
+
+  if (sectrue != norcow_get(PIN_LOGS_KEY, &logs, &len) ||
+      len != PIN_LOG_BYTES) {
+    handle_fault("no PIN logs");
+    return secfalse;
+  }
+
+  uint8_t c = compress_counter(((uint8_t *)logs)[0]);
+
+  secbool success = secfalse;
+
+  uint16_t correct_bytes_cnt = 0;
+
+  for (uint8_t i = 0; i < PIN_LOG_BYTES; i++) {
+    wait_random();
+    correct_bytes_cnt += compress_counter(((uint8_t *)logs)[i]) == c;
+    *ctr = c;
```

Comment:
Is `*ctr = c;` intentionally inside the for cycle?

Reply:
yes i wanted this to be done repeatedly to make potential glitching harder. Though i am not sure this won't get optimized away, i guess nothing really prevents it, so it probably is useless measure